RFCs/0011-partiql-insert-update-spec.md

```diff
@@ -0,0 +1,1041 @@
+* **Item**: A collection of attributes and their corresponding values in a non-relational (E.g. NoSQL) data model.
+* **NoSQL data model**: A non-relational data model that represents the data in a non-tabular way, for example in document or key-value form.
+* **Open schema**: A Schema where undeclared attributes are permitted.
+* **Optional attributes**: A set, comprising the declared attributes that are not necessarily required for a successful data insertion or update to occur.
+* **PartiQL value**: A PartiQL value is a value defined by PartiQL specification that can be one of absent, missing, scalar, tuple, array, or bag. (see PartiQL Specification Figure 1 for more details).
+* **Relation**: A two-dimensional representation of data in rows and columns.
+* **Relational model**: A data model that represents data as relations; with this model, data is organized into relations where each relation is an unordered collection of tuples.
+* **Required attributes**: A set, comprising the declared attributes that must exist for a data item to conform to a Schema.
+* **Schema**: Schema is an abstract entity that is used as a blueprint to describe data along with its associated constraints.
+* **Tuple**: The rows of relations which are collection of column values. In a non-relational data model (E.g. NoSQL/semi-structured model) denotes a collection of name/value pairs also known as an Item; tuples can be ordered or unordered depending on the database system.
+
+### Out of scope
+
+* Returning output from INSERT statement—the specification of returning output is considered as an extension to this RFC which is planned to be defined in the future.
+* Constraints and unique index syntax and semantics—this is considered as a DDL feature as part of `CREATE INDEX` which requires definition in a separate RFC.
+* Conflict detection based on index expression and index predicate—index expression and index predicate grammar and semantics are considered as a DDL feature as part of `CREATE INDEX` which require definition in a separate RFC.
+* Type coercion implementations—the type coercion when inserting data from a source schema to a target schema with an expression (E.g. a sub-select statement) is considered as a database system implementation concern.
+* Security privileges for execution of DML statements—the data access and manipulation security specification is considered as a database system specific feature, hence is considered out of scope for this RFC.
+
+## 2. Proposed grammar and semantics
+
+Figure 1. illustrates the BNF grammar corresponding to PartiQL’s `INSERT INTO ... ON CONFLICT` statements. See Appendix (1) for the Grammar conventions.
+
+Note:
+
+* All the definitions that are missing from the grammar refer to the definitions that are already in SQL-92 standard.
+
+```EBNF
+<insert statement> ::= INSERT INTO <table name> [ AS <alias> ] 
+    [  ( <attr name> [, <attr name> ]... ) ]
+        <values>
```

Comment:
Leaving a post-merge comment here: Since we use `EXCLUDED` as a variable name that references a value to be inserted, we could alternatively propose a follow-up RFC that specifies the syntax to allow an alias for the value to be inserted. Such as:
```ebnf
<insert statement> ::= INSERT INTO <table name> [ AS <alias> ] 
    [  ( <attr name> [, <attr name> ]... ) ]
        <values> [ AS <alias> ]
    ...
```

Note the `[ AS <alias> ]` after `<values>`

Reply:
@johnedquinn it's an intriguing idea: what do you consider the value it brings in addition to `EXCLUDED`?